docs/rfc/rfc-101-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,436 @@
+that is being gossiped is currently valid or not (assuming that invalid transactions could become valid).
+The transaction broadcast logic simply loops through the mempool and tries to send the transactions currently in the pool. 
+
+If we want to ban peers based on duplicate transactions, we should either add additional checks for the cases above, or 
+not ban peers for this behaviour at the moment. It would be useful to gather metrics on how often a peer gossips the same 
+transaction and whether this is cause of significant traffic. 
+
+
+#### **Banning for sending *never-valid* transactions**
+
+If a transaction fails since it could never have been valid, `CheckTx` returns a `ResponseCheckTx.code` 
+value greater than 1. In this case, the peer should be disconnected from and banned immediately without keeping count on how often 
+this has happened. 
+
+The question is whether this transaction should be kept track of in the cache? We can still store it in 
+the cache so that we don't run `CheckTx` on it again, but if this peer is immediately banned, maybe there is no need
+to store its information.
+
+Now, if we want to differentiate further reasons of why this transaction is sent to a node (whether it is a sign of malice or not),
+we might need more information on the actual reason for rejection. This could be done by an additional set of response codes provided by the application. 
+
+### 2. Choosing the peer to ban
+
+Each transaction gossiped contains the ID of the peer that sent that transaction. Upon receiving a transaction,
+a node saves the peer ID of the peer(s) that have sent it. As each peer had to have
+run `CheckTx` on this transaction before adding it to its own mempool, we can assume this peer
+can be held accountable for the validity of transactions it gossips. Invalid transactions are kept only in the mempool
+cache and thus not gossiped. 
+As nodes have to complete a cryptographic handshake at the p2p layer, CometBFT guarantees that a malicious peer
+cannot lie about who the sender of the transaction is. 
+
+*Transactions received from users*
+
+For transactions submitted via `broadcastTxCommit`, the `SenderID` field is empty. 
+
+**Note**  Do we have mechanisms in place to handle cases when `broadcastTxCommit` submits
+failing transactions (can this be a form of attack)?
+
+### 3. Attack scenarios
+
+While an attack by simply banning peers on failing `CheckTx` is hard to imagine, as the incentive for doing so is not clear, there are considerations with regards to the current mempool gossip implementation.
+
+Should we keep transactions that could never have been valid in the cache? Assuming that receiving such transactions is rare, and the peer that sent them is banned, do we need to occupy space in the mempool cache with these transactions?
+
+- What if nodes run different versions of CometBFT and banning is not supported in one of the versions? 
```

Comment:
Being able to ban bad actors is a desirable feature and serves as incentive to upgrade.